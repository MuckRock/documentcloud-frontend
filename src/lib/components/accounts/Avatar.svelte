<script lang="ts">
  import type { User, Org } from "@/api/types";
  import { Person16, Organization16 } from "svelte-octicons";

  export let user: User = undefined;
  export let org: Org = undefined;
</script>

<div class="avatar">
  {#if user}
    {#if user.avatar_url}
      <img src={user.avatar_url} alt="Avatar" />
    {:else}
      <Person16 fill="var(--gray-4)" />
    {/if}
  {:else if org}
    {#if org.avatar_url}
      <img src={org.avatar_url} alt="Avatar" />
    {:else}
      <Organization16 fill="var(--gray-4)" />
    {/if}
  {/if}
</div>

<style>
  .avatar {
    height: 1.5rem;
    width: 1.5rem;
    border-radius: 0.75rem;
    background: var(--gray-2);
    display: flex;
    align-items: center;
    justify-content: center;
    overflow: hidden;
  }

  .avatar > img {
    height: 100%;
    width: 100%;
  }
</style>
