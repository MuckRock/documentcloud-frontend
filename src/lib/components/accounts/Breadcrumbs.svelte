<script lang="ts" context="module">
  export interface Breadcrumb {
    title: string;
    href?: string;
  }
</script>

<script lang="ts">
  import { TriangleRight16 } from "svelte-octicons";
  import Logo from "../common/Logo.svelte";
  import Flex from "../common/Flex.svelte";

  export let trail: Breadcrumb[] = [];
</script>

<div class="breadcrumbs">
  <Flex gap={0.375} align="center">
    <slot name="root">
      <a href="/app" class="logo crumb"><Logo /></a>
    </slot>
    {#each trail as { href, title }}
      <span class="divider"><TriangleRight16 fill="var(--gray-3)" /></span>
      {#if href}
        <a class="crumb" {href} {title}>{title}</a>
      {:else}
        <span {title} class="crumb">{title}</span>
      {/if}
    {/each}
  </Flex>
</div>

<style>
  .breadcrumbs {
    flex: 1 0 0;
    max-width: 100%;
  }

  .crumb {
    flex: 0 1 auto;
    font-size: var(--font-l);
    text-decoration: none;
    color: var(--gray-4);
    padding: 0.25rem 0.5rem;
    border-radius: 0.5rem;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  .crumb:last-child {
    color: inherit;
    flex: 0 0 auto;
  }

  a.crumb:hover {
    background: var(--gray-1);
  }

  .logo {
    flex: 0 0 auto;
    height: 1.5rem;
    width: auto;
    box-sizing: content-box;
  }
</style>
