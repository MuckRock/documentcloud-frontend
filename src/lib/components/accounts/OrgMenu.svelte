<script lang="ts">
  import type { Org } from "@/api/types";
  import { ChevronDown16 } from "svelte-octicons";

  export let org: Org;
</script>

<div class="container">
  <div class="org">
    <div class="avatar">
      <img alt="MuckRock's avatar" src={org.avatar_url} />
    </div>
    <p class="name">{org.name}</p>
    <span class="arrow"><ChevronDown16 /></span>
  </div>
</div>

<style>
  .container {
    width: 100%;
    display: flex;
    gap: 0.5rem;
    flex: 1 0 0;
  }
  .org {
    display: flex;
    align-items: center;
    gap: 0.375rem;
    flex: 1 0 0;
  }
  .avatar {
    width: 1.5rem;
    height: 1.5rem;

    overflow: hidden;
    border-radius: 0.75rem;
    background: var(--gray-2, #d8dee2);
  }
  .avatar img {
    display: block;
    height: 100%;
    width: 100%;
  }
  .name {
    color: #233944;
    font-family: var(--font-sans, "Source Sans Pro");
    font-weight: var(--font-semibold, 600);
    line-height: 1rem;
  }
  .arrow {
    width: 1rem;
    height: 1rem;
    fill: #5c717c;
  }

  /* unused
  .premium {
    display: flex;
    justify-content: center;
    align-items: center;
    flex: 0 0 0;
    align-self: stretch;
    fill: var(--premium);
    height: 2rem;
    width: 2rem;
  } 
  */

  @media (max-width: 64rem) {
    .name {
      display: none;
    }
  }
</style>
