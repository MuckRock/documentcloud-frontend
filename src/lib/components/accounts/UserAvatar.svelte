<script lang="ts">
  import type { User } from "@/api/types";
  import { Person16 } from "svelte-octicons";

  export let user: User = undefined;
</script>

<div class="avatar">
  {#if user?.avatar_url}
    <img src={user.avatar_url} alt="Avatar" />
  {:else}
    <Person16 fill="var(--gray-4)" />
  {/if}
</div>

<style>
  .avatar {
    height: 1.5rem;
    width: 1.5rem;
    border-radius: 0.75rem;
    background: var(--gray-2);
    display: flex;
    align-items: center;
    justify-content: center;
    overflow: hidden;
  }

  .avatar > img {
    height: 100%;
    width: 100%;
  }
</style>
