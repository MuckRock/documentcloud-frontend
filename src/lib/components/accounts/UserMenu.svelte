<script lang="ts">
  import type { User } from "@/api/types";

  import { _ } from "svelte-i18n";
  import { page } from "$app/stores";

  import Dropdown, { closeDropdown } from "@/common/Dropdown2.svelte";
  import SidebarItem from "../sidebar/SidebarItem.svelte";
  import {
    ChevronDown16,
    Gear16,
    Paperclip16,
    SignOut16,
  } from "svelte-octicons";
  import Menu from "@/common/Menu.svelte";

  import { SQUARELET_BASE, SIGN_OUT_URL } from "@/config/config.js";

  export let user: User;

  const dropdownId = "user-menu";
  function close() {
    closeDropdown(dropdownId);
  }
</script>

<Dropdown id={dropdownId} position="right">
  <SidebarItem slot="title">
    <img src={user.avatar_url} alt="Avatar" class="avatar" />
    <span class="name">{user.name}</span>
    <div class="dropdownArrow"><ChevronDown16 /></div>
  </SidebarItem>
  <Menu>
    <SidebarItem href={SQUARELET_BASE} target="_blank" on:click={close}>
      <Gear16 />
      {$_("authSection.user.acctSettings")}
    </SidebarItem>
    <!-- TODO mailkey page -->
    <SidebarItem hover on:click={close}>
      <Paperclip16 />
      {$_("authSection.user.uploadEmail")}
    </SidebarItem>
    <SidebarItem href={SIGN_OUT_URL} on:click={close}>
      <SignOut16 />
      {$_("authSection.user.signOut")}
    </SidebarItem>
  </Menu>
</Dropdown>

<style>
  .avatar {
    height: 1.5rem;
    width: 1.5rem;
    border-radius: 0.75rem;
  }

  .dropdownArrow {
    display: flex;
    align-items: center;
  }

  @media (max-width: 64rem) {
    .name {
      display: none;
    }
  }
</style>
