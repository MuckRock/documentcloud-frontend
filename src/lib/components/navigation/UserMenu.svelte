<script lang="ts">
  import type { User } from "@/api/types";

  import { _ } from "svelte-i18n";

  import Dropdown, { closeDropdown } from "@/common/Dropdown2.svelte";
  import SidebarItem from "../sidebar/SidebarItem.svelte";
  import {
    ChevronDown16,
    Gear16,
    Paperclip16,
    Person16,
    SignOut16,
  } from "svelte-octicons";
  import Menu from "@/common/Menu.svelte";

  import { SQUARELET_BASE, SIGN_OUT_URL } from "@/config/config.js";

  export let user: User;

  const dropdownId = "user-menu";
  function close() {
    closeDropdown(dropdownId);
  }
</script>

<Dropdown id={dropdownId} position="right">
  <SidebarItem slot="title">
    <div class="avatar">
      {#if user.avatar_url}
        <img src={user.avatar_url} alt="Avatar" />
      {:else}
        <Person16 fill="var(--gray-4)" />
      {/if}
    </div>
    <span class="name">{user.name ?? user.username}</span>
    <div class="dropdownArrow"><ChevronDown16 /></div>
  </SidebarItem>
  <Menu>
    <SidebarItem href={SQUARELET_BASE} target="_blank" on:click={close}>
      <Gear16 />
      {$_("authSection.user.acctSettings")}
    </SidebarItem>
    <!-- TODO mailkey page -->
    <SidebarItem hover on:click={close}>
      <Paperclip16 />
      {$_("authSection.user.uploadEmail")}
    </SidebarItem>
    <SidebarItem href={SIGN_OUT_URL} on:click={close}>
      <SignOut16 />
      {$_("authSection.user.signOut")}
    </SidebarItem>
  </Menu>
</Dropdown>

<style>
  .avatar {
    height: 1.5rem;
    width: 1.5rem;
    border-radius: 0.75rem;
    background: var(--gray-2);
    display: flex;
    align-items: center;
    justify-content: center;
    overflow: hidden;
  }

  .avatar > img {
    height: 100%;
    width: 100%;
  }

  .dropdownArrow {
    display: flex;
    align-items: center;
  }

  @media (max-width: 64rem) {
    .name {
      display: none;
    }
  }
</style>
