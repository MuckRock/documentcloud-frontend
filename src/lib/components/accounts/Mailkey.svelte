<script lang="ts">
  import {_} from 'svelte-i18n';
  import Button from "../common/Button.svelte";
  import Flex from "../common/Flex.svelte";
  import Modal from "../layouts/Modal.svelte";
  import { createMailkey, destroyMailkey } from '$lib/api/accounts';

  export let fetch = globalThis.fetch;
  export let message: string = '';
  let error = false;

  function reset() {
    message = '';
    error = false;
  }

  async function create() {
    reset();
    const mailkey = await createMailkey(fetch);
    if (mailkey) {
      message = $_("mailkey.create.success", {
        values: { mailkey: mailkey },
      });
    } else {
      error = true;
      message = $_("mailkey.create.failure")
    }
  }

  async function destroy() {
    reset();
    if (await destroyMailkey(fetch)) {
      message = $_("mailkey.destroy.success");
    } else {
      error = true;
      message = $_("mailkey.destroy.failure");
    }
  }
</script>

<Modal on:close>
  <h1 slot="title">{$_("mailkey.title")}</h1>
  <div class="description">{@html $_("mailkey.description")}</div>
  {#if message}
    <p class:error>
      {@html message}
    </p>
  {/if}
  <Flex gap={1} wrap justify="center">
    <Button mode="primary" on:click={create}>{$_("mailkey.create.button")}</Button>
    <Button mode="danger" on:click={destroy}>{$_("mailkey.destroy.button")}</Button>
  </Flex>
</Modal>

<style>
  .description {
    columns: 2;
    column-gap: 2rem;
    margin-bottom: 1rem;
    font-size: var(--font-sm);
  }
  :global(.description p) {
    margin-bottom: 1rem;
    -webkit-column-break-inside: avoid;
  }
  .error {
    color: var(--red-3);
  }
  @media (max-width: 42rem) {
    .description {
      columns: 1;
    }
  }
</style>