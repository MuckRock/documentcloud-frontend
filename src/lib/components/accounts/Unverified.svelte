<!-- @component
This component displays anywhere we need to explain verification to
a user who is logged in but has `verified_journalist = false`.
-->
<script lang="ts">
  import type { Nullable, User } from "$lib/api/types";

  import { _ } from "svelte-i18n";
  import { Unverified24 } from "svelte-octicons";

  import Button from "../common/Button.svelte";
  import Flex from "../common/Flex.svelte";
  import Tip from "../common/Tip.svelte";

  import {
    APP_URL,
    SQUARELET_ORGS_URL,
    VERIFICATION_FORM_URL,
  } from "@/config/config.js";
  import { isOrg } from "$lib/api/accounts";

  export let user: Nullable<User> = null;

  const FAQ = `${APP_URL}help/faq#verification`;

  function prefill(form_url: string, user: Nullable<User>) {
    if (!user) return form_url;

    const url = new URL(form_url);

    const accountUrl = `https://accounts.muckrock.com/users/${user.username}/`;
    const orgAccountUrl = isOrg(user.organization)
      ? `https://accounts.muckrock.com/organizations/${user.organization.slug}`
      : "";

    const params = {
      "prefill_MR User Email": user.email ?? "",
      "prefill_MR User Name": user.username ?? "",
      "prefill_MR User Account URL": accountUrl,
      "prefill_MR Organization Account URL": orgAccountUrl,
    };

    url.search = new URLSearchParams(params).toString();

    return url.href;
  }
</script>

<Tip>
  <Unverified24 slot="icon" />
  <p>{$_("unverified.verify")}</p>
  <Flex class="buttons" gap={1}>
    <Button mode="primary" href={SQUARELET_ORGS_URL} target="_blank">
      {$_("unverified.orgs")}
    </Button>
    <Button
      mode="primary"
      href={prefill(VERIFICATION_FORM_URL, user)}
      target="_blank"
    >
      {$_("unverified.requestVerificationAction")}
    </Button>
  </Flex>
  <p>{$_("unverified.allowed_actions")}</p>
  <p>
    <a href={FAQ} target="_blank">
      {$_("unverified.faq")}
    </a>
  </p>
</Tip>
