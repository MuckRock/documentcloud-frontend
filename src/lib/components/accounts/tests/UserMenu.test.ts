import { test, expect } from "vitest";
import { render } from "@testing-library/svelte";
import { me as meFixture } from "@/test/fixtures/accounts";
import UserMenu from "../UserMenu.svelte";

test("UserMenu", async () => {
  let result = render(UserMenu, { user: meFixture });
  expect(result.container).toMatchSnapshot();
});
