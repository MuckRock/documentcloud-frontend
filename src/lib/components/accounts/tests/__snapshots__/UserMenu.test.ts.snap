// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserMenu 1`] = `
<div>
   
  <div
    class="anchor svelte-v01npp"
    role="button"
    tabindex="0"
  >
    <span
      class="sidebarItem container svelte-8l7as"
      role="button"
      tabindex="0"
      title="Open Menu"
    >
      <div
        class="avatar svelte-18necmm"
      >
        <img
          alt="Avatar"
          class="svelte-18necmm"
          src="https://cdn.muckrock.com/media/avatars/20140211-0O1A7147-2.jpg"
        />
        
      </div>
       
      <span
        class="label svelte-8l7as"
      >
        <span
          class="name"
        >
          Chris Amico
        </span>
        
      </span>
       
      <div
        class="dropdownArrow svelte-nncahh"
        slot="end"
      >
        <svg
          class="octicon octicon-chevron-down"
          height="12"
          viewBox="0 0 12 12"
          width="12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 8.825c-.2 0-.4-.1-.5-.2l-3.3-3.3c-.3-.3-.3-.8 0-1.1.3-.3.8-.3 1.1 0l2.7 2.7 2.7-2.7c.3-.3.8-.3 1.1 0 .3.3.3.8 0 1.1l-3.2 3.2c-.2.2-.4.3-.6.3Z"
          />
        </svg>
      </div>
    </span>
    
  </div>
   
  <div
    class="dropdown svelte-v01npp"
    style="left: 0px; top: 0px;"
  >
    <div
      class="menu svelte-1tmvwy9"
      role="menu"
    >
      <a
        class="sidebarItem container svelte-8l7as"
        href="https://dev.squarelet.com"
        target="_blank"
        title=""
      >
        <svg
          class="octicon octicon-gear"
          height="16"
          slot="start"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 0a8.2 8.2 0 0 1 .701.031C9.444.095 9.99.645 10.16 1.29l.288 1.107c.018.066.079.158.212.224.231.114.454.243.668.386.123.082.233.09.299.071l1.103-.303c.644-.176 1.392.021 1.82.63.27.385.506.792.704 1.218.315.675.111 1.422-.364 1.891l-.814.806c-.049.048-.098.147-.088.294.016.257.016.515 0 .772-.01.147.038.246.088.294l.814.806c.475.469.679 1.216.364 1.891a7.977 7.977 0 0 1-.704 1.217c-.428.61-1.176.807-1.82.63l-1.102-.302c-.067-.019-.177-.011-.3.071a5.909 5.909 0 0 1-.668.386c-.133.066-.194.158-.211.224l-.29 1.106c-.168.646-.715 1.196-1.458 1.26a8.006 8.006 0 0 1-1.402 0c-.743-.064-1.289-.614-1.458-1.26l-.289-1.106c-.018-.066-.079-.158-.212-.224a5.738 5.738 0 0 1-.668-.386c-.123-.082-.233-.09-.299-.071l-1.103.303c-.644.176-1.392-.021-1.82-.63a8.12 8.12 0 0 1-.704-1.218c-.315-.675-.111-1.422.363-1.891l.815-.806c.05-.048.098-.147.088-.294a6.214 6.214 0 0 1 0-.772c.01-.147-.038-.246-.088-.294l-.815-.806C.635 6.045.431 5.298.746 4.623a7.92 7.92 0 0 1 .704-1.217c.428-.61 1.176-.807 1.82-.63l1.102.302c.067.019.177.011.3-.071.214-.143.437-.272.668-.386.133-.066.194-.158.211-.224l.29-1.106C6.009.645 6.556.095 7.299.03 7.53.01 7.764 0 8 0Zm-.571 1.525c-.036.003-.108.036-.137.146l-.289 1.105c-.147.561-.549.967-.998 1.189-.173.086-.34.183-.5.29-.417.278-.97.423-1.529.27l-1.103-.303c-.109-.03-.175.016-.195.045-.22.312-.412.644-.573.99-.014.031-.021.11.059.19l.815.806c.411.406.562.957.53 1.456a4.709 4.709 0 0 0 0 .582c.032.499-.119 1.05-.53 1.456l-.815.806c-.081.08-.073.159-.059.19.162.346.353.677.573.989.02.03.085.076.195.046l1.102-.303c.56-.153 1.113-.008 1.53.27.161.107.328.204.501.29.447.222.85.629.997 1.189l.289 1.105c.029.109.101.143.137.146a6.6 6.6 0 0 0 1.142 0c.036-.003.108-.036.137-.146l.289-1.105c.147-.561.549-.967.998-1.189.173-.086.34-.183.5-.29.417-.278.97-.423 1.529-.27l1.103.303c.109.029.175-.016.195-.045.22-.313.411-.644.573-.99.014-.031.021-.11-.059-.19l-.815-.806c-.411-.406-.562-.957-.53-1.456a4.709 4.709 0 0 0 0-.582c-.032-.499.119-1.05.53-1.456l.815-.806c.081-.08.073-.159.059-.19a6.464 6.464 0 0 0-.573-.989c-.02-.03-.085-.076-.195-.046l-1.102.303c-.56.153-1.113.008-1.53-.27a4.44 4.44 0 0 0-.501-.29c-.447-.222-.85-.629-.997-1.189l-.289-1.105c-.029-.11-.101-.143-.137-.146a6.6 6.6 0 0 0-1.142 0ZM11 8a3 3 0 1 1-6 0 3 3 0 0 1 6 0ZM9.5 8a1.5 1.5 0 1 0-3.001.001A1.5 1.5 0 0 0 9.5 8Z"
          />
        </svg>
         
        <span
          class="label svelte-8l7as"
        >
          Account settings
        </span>
         
      </a>
      
       
      <span
        class="sidebarItem container svelte-8l7as hover"
        role="button"
        tabindex="0"
        title=""
      >
        <svg
          class="octicon octicon-paperclip"
          height="16"
          slot="start"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.212 3.02a1.753 1.753 0 0 0-2.478.003l-5.83 5.83a3.007 3.007 0 0 0-.88 2.127c0 .795.315 1.551.88 2.116.567.567 1.333.89 2.126.89.79 0 1.548-.321 2.116-.89l5.48-5.48a.75.75 0 0 1 1.061 1.06l-5.48 5.48a4.492 4.492 0 0 1-3.177 1.33c-1.2 0-2.345-.487-3.187-1.33a4.483 4.483 0 0 1-1.32-3.177c0-1.195.475-2.341 1.32-3.186l5.83-5.83a3.25 3.25 0 0 1 5.553 2.297c0 .863-.343 1.691-.953 2.301L7.439 12.39c-.375.377-.884.59-1.416.593a1.998 1.998 0 0 1-1.412-.593 1.992 1.992 0 0 1 0-2.828l5.48-5.48a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-5.48 5.48a.492.492 0 0 0 0 .707.499.499 0 0 0 .352.154.51.51 0 0 0 .356-.154l5.833-5.827a1.755 1.755 0 0 0 0-2.481Z"
          />
        </svg>
         
        <span
          class="label svelte-8l7as"
        >
          Upload via email
        </span>
         
      </span>
      
       
      <a
        class="sidebarItem container svelte-8l7as"
        href="https://api.dev.documentcloud.org/accounts/logout/"
        title=""
      >
        <svg
          class="octicon octicon-sign-out"
          height="16"
          slot="start"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2 2.75C2 1.784 2.784 1 3.75 1h2.5a.75.75 0 0 1 0 1.5h-2.5a.25.25 0 0 0-.25.25v10.5c0 .138.112.25.25.25h2.5a.75.75 0 0 1 0 1.5h-2.5A1.75 1.75 0 0 1 2 13.25Zm10.44 4.5-1.97-1.97a.749.749 0 0 1 .326-1.275.749.749 0 0 1 .734.215l3.25 3.25a.75.75 0 0 1 0 1.06l-3.25 3.25a.749.749 0 0 1-1.275-.326.749.749 0 0 1 .215-.734l1.97-1.97H6.75a.75.75 0 0 1 0-1.5Z"
          />
        </svg>
         
        <span
          class="label svelte-8l7as"
        >
          Sign out
        </span>
         
      </a>
      
    </div>
  </div>
   
  
</div>
`;
